#Original Author: Jonathan Hudson
#CPSC 501 F22

In [2]:
import tensorflow as tf

tf.random.set_seed(1234)

print("--Get data--")
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

print("--Process data--")
x_train, x_test = x_train / 255.0, x_test / 255.0

2022-11-11 13:39:35.470430: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-11-11 13:39:35.573529: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-11-11 13:39:35.573543: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2022-11-11 13:39:35.595296: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2022-11-11 13:39:36.670773: W tensorflow/stream_executor/platform/de

--Get data--
--Process data--


In [8]:
print("--Make model--")
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax'),
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

print("--Fit model--")
model.fit(x_train, y_train, epochs=30, verbose=2)

--Make model--
--Fit model--
Epoch 1/30
1875/1875 - 2s - loss: 0.2360 - accuracy: 0.9289 - 2s/epoch - 1ms/step
Epoch 2/30
1875/1875 - 2s - loss: 0.1032 - accuracy: 0.9685 - 2s/epoch - 1ms/step
Epoch 3/30
1875/1875 - 2s - loss: 0.0775 - accuracy: 0.9766 - 2s/epoch - 1ms/step
Epoch 4/30
1875/1875 - 2s - loss: 0.0622 - accuracy: 0.9813 - 2s/epoch - 1ms/step
Epoch 5/30
1875/1875 - 2s - loss: 0.0495 - accuracy: 0.9842 - 2s/epoch - 1ms/step
Epoch 6/30
1875/1875 - 2s - loss: 0.0424 - accuracy: 0.9870 - 2s/epoch - 1ms/step
Epoch 7/30
1875/1875 - 2s - loss: 0.0395 - accuracy: 0.9877 - 2s/epoch - 1ms/step
Epoch 8/30
1875/1875 - 2s - loss: 0.0316 - accuracy: 0.9900 - 2s/epoch - 1ms/step
Epoch 9/30
1875/1875 - 2s - loss: 0.0293 - accuracy: 0.9907 - 2s/epoch - 1ms/step
Epoch 10/30
1875/1875 - 2s - loss: 0.0278 - accuracy: 0.9911 - 2s/epoch - 1ms/step
Epoch 11/30
1875/1875 - 2s - loss: 0.0237 - accuracy: 0.9926 - 2s/epoch - 1ms/step
Epoch 12/30
1875/1875 - 2s - loss: 0.0229 - accuracy: 0.9931 - 2s/e

In [11]:
print("--Evaluate model--")
model_loss1, model_acc1 = model.evaluate(x_train,  y_train, verbose=2)
model_loss2, model_acc2 = model.evaluate(x_test,  y_test, verbose=2)
print(f"Train / Test Accuracy: {model_acc1*100:.1f}% / {model_acc2*100:.1f}%")
model.save("mnist_98.1_accuracy", save_format="h5") #Completed model with Train / Test Accuracy: 99.8% / 98.1%

--Evaluate model--
1875/1875 - 1s - loss: 0.0069 - accuracy: 0.9981 - 1s/epoch - 735us/step
313/313 - 0s - loss: 0.1330 - accuracy: 0.9807 - 249ms/epoch - 796us/step
Train / Test Accuracy: 99.8% / 98.1%



Report
--
Getting the accuracy to 98+% was much easier than I had initially expected with this dataset. The only modification to hyper parameters that I made, was that I added more Rectified linear unit activation (RELU) layers to my model, and I also increased the density of the layers. Within these models there tends to be a lot of data that needs to be normalized, and if there is nodes that represent probabilities, we dont want these values to be non-positive. The RELU collapses all these negative values to 0, and I found that adding one got my accuracy to increase, and adding another got it to increase more. After adding a few more I was worried about overfitting, so I stopped adding them. But this did make the model get more out of the data that it was given, which is an improvement.
Changes:
```py
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax'),
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

```

When looking for other ways to improve the starter code, I noticed that the model was only trained for 1 epoch. I decided that this was less than ideal, and made it train for 30 epochs, which was still relatively quick. I found that after 30, there was quite a bit of diminishing returns in the amount of accuracy that it improved and the amount of time that it took for the model to reach new milestones in accuracy. When giving it higher epochs to train for, I found that this caused the model to overfit to the dataset.

Another change that I made was that I changed the optimizer from SGD to Adam, because when looking through the documentation I found that Adam tends to have better computation time, and requires fewer hyper parameters for fine tuning.
Changes:
```py
model.fit(x_train, y_train, epochs=30, verbose=2)
```
Overall, these three simple changes were more than enough to cause the model to blow past the 95% threshold that was required, and hit the 98% for bonus. I think the largest flaw with the model to begin with was that it wasnt normalizing the data that it had, and it also wasnt given enough time to make mistakes and improve on those mistakes with only 1 epoch. I think a basic model such as this, needs at least 30 epochs to be able to analyze data effectively with a large enough dataset. So by fixing those two mistakes, the model is in a much better position.



